EventFilter/DTRawToDigi/plugins/DTuROSDigiToRaw.cc

```diff
@@ -0,0 +1,436 @@
+    LineFED+=4;
+    *((int*)LineFED)=dataWord1; 
+  }
+
+  //--> AMC data
+
+  for (int sltit = 1; sltit <= 12; sltit++) {
+
+    if (bslts[sltit] == 0) continue;
+
+    dataWord1 = (sltit&0xF)<<24;
+    dataWord2 = 0;
+
+    calcCRC(dataWord1, dataWord2, newCRC);
+    LineFED+=4;
+    *((int*)LineFED)=dataWord2; 
+    LineFED+=4;
+    *((int*)LineFED)=dataWord1; 
+
+    dataWord1 = 0;
+    dataWord2 = 0;
+
+    calcCRC(dataWord1, dataWord2, newCRC);
+    LineFED+=4;
+    *((int*)LineFED)=dataWord2; 
+    LineFED+=4;
+    *((int*)LineFED)=dataWord1; 
+
+    for (int nhit = 0; nhit < bslts[sltit]/2; nhit++) {
+
+      dataWord1 = 0x20000000
+                + wslts[sltit].at(nhit*2);
+      dataWord2 = wslts[sltit].at(nhit*2+1);
+
+      calcCRC(dataWord1, dataWord2, newCRC);
+      LineFED+=4;
+      *((int*)LineFED)=dataWord2; 
+      LineFED+=4;
+      *((int*)LineFED)=dataWord1; 
+    }
+
+    if (bslts[sltit]%2 == 1) {
+
+      dataWord1 = 0x20000000
+                + wslts[sltit].at(bslts[sltit]-1);
+      dataWord2 = 0;
+
+      calcCRC(dataWord1, dataWord2, newCRC);
+      LineFED+=4;
+      *((int*)LineFED)=dataWord2; 
+      LineFED+=4;
+      *((int*)LineFED)=dataWord1; 
+    }
+
+    dataWord1 = 0x40000000;
+    dataWord2 = 0;
+
+    calcCRC(dataWord1, dataWord2, newCRC);
+    LineFED+=4;
+    *((int*)LineFED)=dataWord2; 
+    LineFED+=4;
+    *((int*)LineFED)=dataWord1; 
+
+    dataWord1 = 0x40000000;
+    dataWord2 = 0;
+
+    calcCRC(dataWord1, dataWord2, newCRC);
+    LineFED+=4;
+    *((int*)LineFED)=dataWord2; 
+    LineFED+=4;
+    *((int*)LineFED)=dataWord1; 
+
+    dataWord1 = 0;
+    dataWord2 = (dslts[sltit]&0xFFFFF);
+
+    calcCRC(dataWord1, dataWord2, newCRC);
+    LineFED+=4;
+    *((int*)LineFED)=dataWord2; 
+    LineFED+=4;
+    *((int*)LineFED)=dataWord1; 
+  }
+
+  //--> Trailer - line 1
+
+  dataWord1 = 0;
+  dataWord2 = 0;
+
+  calcCRC(dataWord1, dataWord2, newCRC);
+  LineFED+=4;
+  *((int*)LineFED)=dataWord2; 
+  LineFED+=4;
+  *((int*)LineFED)=dataWord1; 
+
+  //--> Trailer - line 2
+
+  dataWord1 = 0xA0000000
+             + (lines&0xFFFFFF);
+  dataWord2 = 0;
+
+  calcCRC(dataWord1, dataWord2&0xFFFF, newCRC);
+
+  dataWord2 += (newCRC&0xFFFF)<<16;
+
+  LineFED+=4;
+  *((int*)LineFED)=dataWord2; 
+  LineFED+=4;
+  *((int*)LineFED)=dataWord1; 
+
+  return;
+}
+
+
+void DTuROSDigiToRaw::clear() {
+
+  for (int sltit = 0; sltit <= 12; sltit++) {
+
+    bslts[sltit]=0;
+    dslts[sltit]=0;
+    wslts[sltit].clear();
+  }
+
+  return;
+}
+
+
+int DTuROSDigiToRaw::theCRT(int ddu, int ros, int rob) {
+
+  if (ddu == 770) return 1368;
+  if (ddu == 771) return 1368;
+  if (ddu == 772) return 1369;
+  return 1370;
+}
+
+
+int DTuROSDigiToRaw::theSLT(int ddu, int ros, int rob) {
+
+  int slot = ((ros-1)/3)+1;
+  if (rob == 23) slot = 5;
+  if (ddu == 771) slot += 6;
+  if (ddu == 773) slot += 6;
+  return slot;
+}
+
+
+int DTuROSDigiToRaw::theLNK(int ddu, int ros, int rob) {
+
+  int link = rob;
```

Comment:
- Possible simplification of the code:

int link = (rob>14? rob+1:rob);
if (rob == 24) link = 15;
if (rob == 23) link = ros-1;
else link += ((ros-1)%3)*24;
return link;

Reply:
don't want to mix these two (very different) conditions